specs/capri/API_Framework_Refactoring.md

```diff
@@ -0,0 +1,384 @@
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// CreateVolumeAttachmentOpts is a structure which indicates all required properties
+// for deleting a volume attachment.
+message DeleteVolumeAttachmentOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// CreateVolumeSnapshotOpts is a structure which indicates all required properties
+// for creating a volume snapshot.
+message CreateVolumeSnapshotOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// DeleteVolumeSnapshotOpts is a structure which indicates all required properties
+// for deleting a volume snapshot.
+message DeleteVolumeSnapshotOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// CreateReplicationOpts is a structure which indicates all required properties
+// for creating a replication.
+// NOTE: Need to figure out how to handle more than 2 sites.
+message CreateReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+// for deleting a replication.
+// NOTE: Need to figure out how to handle more than 2 sites.
+message DeleteReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+message EnableReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+message DisableReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+message FailoverReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    string failoverMessage = 2;
+    // The Context
+    string context = 3;
+}
+
+// CreateVolumeGroupOpts is a structure which indicates all required
+// properties for creating a volume group.
+message CreateVolumeGroupOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+message UpdateVolumeGroupOpts{
+    // The message of request, required.
+    string message = 1;
+    repeated string addVolMessage = 2;
+    repeated string rmVolMessage = 3;
+    // The Context
+    string context = 4;
+}
+
+message DeleteVolumeGroupOpts{
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+message GenericResponse {}
+```
```

Comment:
We should have Get and List methods for each resource to check the status of an operation.

Reply:
One question for these methods, since we can check the status of operation through etcd, is there any need to double-check the status by calling the backend?

Reply:
How about adding a QoS parameter into the design like some messaging implementations.
Level 1: no check
Level 2: check backend.

Reply:
@noelmcloughlin Good suggestion, but actually in the long term we want to redesign a unified model for both controller and dock module.